File: edge/app/backup/archiver.py
from __future__ import annotations

import os
import tarfile
import tempfile
from datetime import datetime, timezone
from pathlib import Path

import zstandard

from app.backup.filters import DiscoveredFile


def timestamp_for_filename(timestamp: datetime) -> str:
    return timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H-%M-%SZ")


def timestamp_for_api(timestamp: datetime) -> str:
    return timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def build_archive_name(job_name: str, timestamp: datetime, fingerprint: str) -> str:
    return f"{job_name}__{timestamp_for_filename(timestamp)}__{fingerprint[:8]}.tar.zst"


def create_archive(archive_path: Path, files: list[DiscoveredFile]) -> None:
    archive_path.parent.mkdir(parents=True, exist_ok=True)
    compressor = zstandard.ZstdCompressor(level=3)

    with archive_path.open("wb") as raw_handle:
        # Keep the underlying file handle open so we can flush and fsync it
        # after the streaming tar+zstd write completes.
        with compressor.stream_writer(raw_handle, closefd=False) as compressed_handle:
            with tarfile.open(mode="w|", fileobj=compressed_handle, format=tarfile.PAX_FORMAT) as tar_handle:
                for file in sorted(files, key=lambda item: item.archive_path):
                    tar_info = tarfile.TarInfo(name=file.archive_path)
                    tar_info.size = file.size
                    tar_info.mtime = file.mtime_ns / 1_000_000_000
                    tar_info.mode = 0o644
                    tar_info.uid = 0
                    tar_info.gid = 0
                    tar_info.uname = ""
                    tar_info.gname = ""

                    with file.source_path.open("rb") as source_handle:
                        tar_handle.addfile(tar_info, source_handle)

        raw_handle.flush()
        os.fsync(raw_handle.fileno())


def _archive_destination(target_root: Path, member_name: str) -> Path:
    destination = (target_root / member_name).resolve()
    try:
        destination.relative_to(target_root)
    except ValueError as exc:
        raise ValueError(f"invalid archive entry: {member_name}") from exc
    return destination


def list_archive_entries(archive_path: Path, target_root: Path) -> dict[str, object]:
    target_root = target_root.resolve()
    entries: list[dict[str, object]] = []
    replace_count = 0
    add_count = 0
    decompressor = zstandard.ZstdDecompressor()

    with archive_path.open("rb") as raw_handle:
        with decompressor.stream_reader(raw_handle) as decompressed_handle:
            with tarfile.open(mode="r|", fileobj=decompressed_handle) as tar_handle:
                for member in tar_handle:
                    if member.isdir():
                        continue
                    if not member.isfile():
                        raise ValueError(f"unsupported archive entry: {member.name}")

                    destination = _archive_destination(target_root, member.name)
                    action = "replace" if destination.exists() else "add"
                    if action == "replace":
                        replace_count += 1
                    else:
                        add_count += 1
                    entries.append(
                        {
                            "path": member.name,
                            "size": member.size,
                            "mtime": member.mtime,
                            "action": action,
                        }
                    )

    return {
        "entries": entries,
        "total_files": len(entries),
        "replace_count": replace_count,
        "add_count": add_count,
    }


def extract_archive(archive_path: Path, target_root: Path) -> int:
    target_root = target_root.resolve()
    extracted_count = 0
    decompressor = zstandard.ZstdDecompressor()

    with archive_path.open("rb") as raw_handle:
        with decompressor.stream_reader(raw_handle) as decompressed_handle:
            with tarfile.open(mode="r|", fileobj=decompressed_handle) as tar_handle:
                for member in tar_handle:
                    if member.isdir():
                        continue
                    if not member.isfile():
                        raise ValueError(f"unsupported archive entry: {member.name}")

                    destination = _archive_destination(target_root, member.name)
                    source_handle = tar_handle.extractfile(member)
                    if source_handle is None:
                        raise ValueError(f"unable to extract archive entry: {member.name}")

                    destination.parent.mkdir(parents=True, exist_ok=True)
                    with tempfile.NamedTemporaryFile(
                        mode="wb",
                        dir=destination.parent,
                        delete=False,
                        suffix=".restore.tmp",
                    ) as temp_handle:
                        temp_path = Path(temp_handle.name)
                        try:
                            while True:
                                chunk = source_handle.read(1024 * 1024)
                                if not chunk:
                                    break
                                temp_handle.write(chunk)
                            temp_handle.flush()
                            os.fsync(temp_handle.fileno())
                        finally:
                            source_handle.close()

                    try:
                        os.replace(temp_path, destination)
                        os.utime(destination, (member.mtime, member.mtime))
                    finally:
                        temp_path.unlink(missing_ok=True)
                    extracted_count += 1

    return extracted_count

File: edge/internal/backup/archiver.go
package backup

import (
	"archive/tar"
	"fmt"
	"io"
	"os"
	"cmp"
	"path/filepath"
	"slices"
	"strings"
	"time"

	securejoin "github.com/cyphar/filepath-securejoin"
	"github.com/klauspost/compress/zstd"
)

// TimestampForFilename formats a UTC time for use in archive filenames.
func TimestampForFilename(t time.Time) string {
	return t.UTC().Format("2006-01-02T15-04-05Z")
}

// TimestampForAPI formats a UTC time for API payloads.
func TimestampForAPI(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05Z")
}

// BuildArchiveName returns the canonical archive filename.
func BuildArchiveName(jobName string, t time.Time, fingerprint string) string {
	fp := fingerprint
	if len(fp) > 8 {
		fp = fp[:8]
	}
	return fmt.Sprintf("%s__%s__%s.tar.zst", jobName, TimestampForFilename(t), fp)
}

// CreateArchive writes a zstd-compressed PAX tar of files to archivePath.
func CreateArchive(archivePath string, files []*DiscoveredFile) error {
	if err := os.MkdirAll(filepath.Dir(archivePath), 0o755); err != nil {
		return err
	}

	sorted := make([]*DiscoveredFile, len(files))
	copy(sorted, files)
	slices.SortFunc(sorted, func(a, b *DiscoveredFile) int {
		return cmp.Compare(a.ArchivePath, b.ArchivePath)
	})

	f, err := os.Create(archivePath)
	if err != nil {
		return err
	}
	defer f.Close()

	enc, err := zstd.NewWriter(f, zstd.WithEncoderLevel(zstd.SpeedDefault))
	if err != nil {
		return err
	}

	tw := tar.NewWriter(enc)
	for _, file := range sorted {
		if err := addFileToTar(tw, file); err != nil {
			enc.Close()
			return err
		}
	}
	if err := tw.Close(); err != nil {
		enc.Close()
		return err
	}
	if err := enc.Close(); err != nil {
		return err
	}
	if err := f.Sync(); err != nil {
		return err
	}
	return nil
}

func addFileToTar(tw *tar.Writer, file *DiscoveredFile) error {
	src, err := os.Open(file.SourcePath)
	if err != nil {
		return err
	}
	defer src.Close()

	hdr := &tar.Header{
		Name:    file.ArchivePath,
		Size:    file.Size,
		ModTime: time.Unix(0, file.MtimeNs),
		Mode:    0o644,
		Uid:     0, Gid: 0,
		Uname:  "",
		Gname:  "",
		Format: tar.FormatPAX,
	}
	if err := tw.WriteHeader(hdr); err != nil {
		return err
	}
	_, err = io.Copy(tw, src)
	return err
}

// ListArchiveEntries streams the zstd+tar archive and returns a summary of its contents.
func ListArchiveEntries(archivePath, targetRoot string) (map[string]interface{}, error) {
	absTarget, err := filepath.Abs(targetRoot)
	if err != nil {
		return nil, err
	}

	f, err := os.Open(archivePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	dec, err := zstd.NewReader(f)
	if err != nil {
		return nil, err
	}
	defer dec.Close()

	type entry struct {
		Path   string  `json:"path"`
		Size   int64   `json:"size"`
		Mtime  float64 `json:"mtime"`
		Action string  `json:"action"`
	}

	var entries []entry
	replaceCount, addCount := 0, 0

	tr := tar.NewReader(dec)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if hdr.Typeflag == tar.TypeDir {
			continue
		}
		if hdr.Typeflag != tar.TypeReg {
			return nil, fmt.Errorf("unsupported archive entry: %s", hdr.Name)
		}
		dest, err := archiveDestination(absTarget, hdr.Name)
		if err != nil {
			return nil, err
		}
		action := "add"
		if _, err := os.Stat(dest); err == nil {
			action = "replace"
			replaceCount++
		} else {
			addCount++
		}
		entries = append(entries, entry{
			Path:   hdr.Name,
			Size:   hdr.Size,
			Mtime:  float64(hdr.ModTime.Unix()),
			Action: action,
		})
	}

	result := make([]interface{}, len(entries))
	for i, e := range entries {
		result[i] = map[string]interface{}{
			"path":   e.Path,
			"size":   e.Size,
			"mtime":  e.Mtime,
			"action": e.Action,
		}
	}
	return map[string]interface{}{
		"entries":       result,
		"total_files":   len(entries),
		"replace_count": replaceCount,
		"add_count":     addCount,
	}, nil
}

// ExtractArchive streams the zstd+tar archive and writes files to targetRoot.
func ExtractArchive(archivePath, targetRoot string) (int, error) {
	absTarget, err := filepath.Abs(targetRoot)
	if err != nil {
		return 0, err
	}

	f, err := os.Open(archivePath)
	if err != nil {
		return 0, err
	}
	defer f.Close()

	dec, err := zstd.NewReader(f)
	if err != nil {
		return 0, err
	}
	defer dec.Close()

	extractedCount := 0
	tr := tar.NewReader(dec)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return extractedCount, err
		}
		if hdr.Typeflag == tar.TypeDir {
			continue
		}
		if hdr.Typeflag != tar.TypeReg {
			return extractedCount, fmt.Errorf("unsupported archive entry: %s", hdr.Name)
		}

		dest, err := archiveDestination(absTarget, hdr.Name)
		if err != nil {
			return extractedCount, err
		}
		if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
			return extractedCount, err
		}

		tmp := dest + ".restore.tmp"
		if err := writeAtomic(tmp, dest, hdr.ModTime, tr); err != nil {
			os.Remove(tmp)
			return extractedCount, err
		}
		extractedCount++
	}
	return extractedCount, nil
}

func writeAtomic(tmp, dest string, mtime time.Time, r io.Reader) error {
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o644)
	if err != nil {
		return err
	}
	if _, err := io.Copy(f, r); err != nil {
		f.Close()
		return err
	}
	if err := f.Sync(); err != nil {
		f.Close()
		return err
	}
	f.Close()
	if err := os.Rename(tmp, dest); err != nil {
		return err
	}
	return os.Chtimes(dest, mtime, mtime)
}

func archiveDestination(targetRoot, memberName string) (string, error) {
	cleaned := filepath.Clean(memberName)
	if filepath.IsAbs(cleaned) || strings.HasPrefix(cleaned, "..") {
		return "", fmt.Errorf("invalid archive entry: %s", memberName)
	}
	dest, err := securejoin.SecureJoin(targetRoot, cleaned)
	if err != nil {
		return "", fmt.Errorf("invalid archive entry: %s", memberName)
	}
	return dest, nil
}
